lib/ansible/modules/crypto/openssl_certificate.py

```diff
@@ -786,11 +1042,431 @@ def dump(self, check_mode=False):
         return result
 
 
+class AssertOnlyCertificateCryptography(Certificate):
+    """Validate the supplied cert, using the cryptography backend"""
+    def __init__(self, module):
+        super(AssertOnlyCertificateCryptography, self).__init__(module, 'cryptography')
+        self.signature_algorithms = module.params['signature_algorithms']
+        if module.params['subject']:
+            self.subject = crypto_utils.parse_name_field(module.params['subject'])
+        else:
+            self.subject = []
+        self.subject_strict = module.params['subject_strict']
+        if module.params['issuer']:
+            self.issuer = crypto_utils.parse_name_field(module.params['issuer'])
+        else:
+            self.issuer = []
+        self.issuer_strict = module.params['issuer_strict']
+        self.has_expired = module.params['has_expired']
+        self.version = module.params['version']
+        self.keyUsage = module.params['key_usage']
+        self.keyUsage_strict = module.params['key_usage_strict']
+        self.extendedKeyUsage = module.params['extended_key_usage']
+        self.extendedKeyUsage_strict = module.params['extended_key_usage_strict']
+        self.subjectAltName = module.params['subject_alt_name']
+        self.subjectAltName_strict = module.params['subject_alt_name_strict']
+        self.notBefore = module.params['not_before'],
+        self.notAfter = module.params['not_after'],
+        self.valid_at = module.params['valid_at'],
+        self.invalid_at = module.params['invalid_at'],
+        self.valid_in = module.params['valid_in'],
+        self.message = []
+        # self._sanitize_inputs()
+
+    def _sanitize_inputs(self):
+        """Ensure inputs are properly sanitized before comparison."""
+
+        for param in ['signature_algorithms', 'keyUsage', 'extendedKeyUsage',
+                      'subjectAltName', 'subject', 'issuer', 'notBefore',
+                      'notAfter', 'valid_at', 'invalid_at']:
+
+            attr = getattr(self, param)
+            if isinstance(attr, list) and attr:
+                if isinstance(attr[0], str):
+                    setattr(self, param, [to_bytes(item) for item in attr])
+                elif isinstance(attr[0], tuple):
+                    setattr(self, param, [(to_bytes(item[0]), to_bytes(item[1])) for item in attr])
+            elif isinstance(attr, tuple):
+                setattr(self, param, dict((to_bytes(k), to_bytes(v)) for (k, v) in attr.items()))
+                # setattr(self, param, dict((to_bytes(k), to_bytes(v)) for (k, v) in attr))
+            elif isinstance(attr, dict):
+                setattr(self, param, dict((to_bytes(k), to_bytes(v)) for (k, v) in attr.items()))
+                # setattr(self, param, dict((to_bytes(k), to_bytes(v)) for (k, v) in attr))
+            elif isinstance(attr, str):
+                setattr(self, param, to_bytes(attr))
+
+    def _get_name_oid(self, id):
+        if id in ('CN', 'commonName'):
```

Comment:
Let's move these helper functions to module_utils and make openssl_csr and openssl_certificate use the same functions. But let's not do it before merging this PR, otherwise we lose too much time :)

Reply:
Yep, sounds like a plan